File: Controllers/CartController.vb
Imports System.Web.Mvc
Imports Newtonsoft.Json
Imports My.Models

Namespace My.Controllers
    Public Class CartController
        Inherits Controller

        Private Const CartSessionKey As String = "ShoppingCart"

        Private Function GetCart() As CartViewModel
            Dim cart = TryCast(Session(CartSessionKey), CartViewModel)
            If cart Is Nothing Then
                cart = New CartViewModel With {.Items = New List(Of CartItem)()}
                Session(CartSessionKey) = cart
            End If
            Return cart
        End Function

        ' GET: /Cart
        Function Index() As ActionResult
            Return View(GetCart())
        End Function

        ' POST: /Cart/Add
        <HttpPost>
        <ValidateAntiForgeryToken>
        Function Add(productId As Integer, qty As Integer) As ActionResult
            ' TODO: 查詢商品資訊
            Dim cart = GetCart()
            Dim existing = cart.Items.FirstOrDefault(Function(x) x.ProductId = productId)
            If existing IsNot Nothing Then
                existing.Quantity += qty
            Else
                ' cart.Items.Add(New CartItem With { ... })
            End If
            Session(CartSessionKey) = cart
            Return Json(New With {.success = True, .totalQty = cart.TotalQty})
        End Function

        ' POST: /Cart/Update
        <HttpPost>
        <ValidateAntiForgeryToken>
        Function Update(productId As Integer, qty As Integer) As ActionResult
            Dim cart = GetCart()
            Dim item = cart.Items.FirstOrDefault(Function(x) x.ProductId = productId)
            If item IsNot Nothing Then
                If qty <= 0 Then
                    cart.Items.Remove(item)
                Else
                    item.Quantity = qty
                End If
            End If
            Session(CartSessionKey) = cart
            Return RedirectToAction("Index")
        End Function

        ' POST: /Cart/Remove
        <HttpPost>
        <ValidateAntiForgeryToken>
        Function Remove(productId As Integer) As ActionResult
            Dim cart = GetCart()
            cart.Items.RemoveAll(Function(x) x.ProductId = productId)
            Session(CartSessionKey) = cart
            Return RedirectToAction("Index")
        End Function

        ' GET: /Cart/Checkout
        <Authorize>
        Function Checkout() As ActionResult
            Dim cart = GetCart()
            If cart.TotalQty = 0 Then Return RedirectToAction("Index")
            Dim vm = New CheckoutViewModel With {.Cart = cart}
            Return View(vm)
        End Function

        ' POST: /Cart/Checkout
        <HttpPost>
        <Authorize>
        <ValidateAntiForgeryToken>
        Function Checkout(model As CheckoutViewModel) As ActionResult
            If Not ModelState.IsValid Then
                model.Cart = GetCart()
                Return View(model)
            End If
            ' TODO: 建立訂單、清空購物車
            Session.Remove(CartSessionKey)
            Return RedirectToAction("Complete")
        End Function

        ' GET: /Cart/Complete
        Function Complete() As ActionResult
            Return View()
        End Function

    End Class
End Namespace

File: Controllers/CartController.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using MyWeb.Models;

namespace MyWeb.Controllers
{
    public class CartController : Controller
    {
        private const string CartSessionKey = "ShoppingCart";

        private CartViewModel GetCart()
        {
            var cart = Session[CartSessionKey] as CartViewModel;
            if (cart == null)
            {
                cart = new CartViewModel { Items = new List<CartItem>() };
                Session[CartSessionKey] = cart;
            }
            return cart;
        }

        // GET: /Cart
        public ActionResult Index()
        {
            return View(GetCart());
        }

        // POST: /Cart/Add
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Add(int productId, int qty)
        {
            if (qty < 1) qty = 1;

            using (var db = new AppDbContext())
            {
                var product = db.Products.FirstOrDefault(p => p.Id == productId && p.IsActive);
                if (product == null)
                    return Json(new { success = false, message = "商品不存在" });

                var cart = GetCart();
                var existing = cart.Items.FirstOrDefault(x => x.ProductId == productId);
                if (existing != null)
                {
                    existing.Quantity += qty;
                }
                else
                {
                    cart.Items.Add(new CartItem
                    {
                        ProductId = product.Id,
                        ProductName = product.Name,
                        ImagePath = product.ImagePath,
                        UnitPrice = product.SalePrice ?? product.Price,
                        Quantity = qty
                    });
                }
                Session[CartSessionKey] = cart;
                return Json(new { success = true, totalQty = cart.TotalQty, productName = product.Name });
            }
        }

        // POST: /Cart/Update
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Update(int productId, int qty)
        {
            var cart = GetCart();
            var item = cart.Items.FirstOrDefault(x => x.ProductId == productId);
            if (item != null)
            {
                if (qty <= 0)
                {
                    cart.Items.Remove(item);
                }
                else
                {
                    item.Quantity = qty;
                }
            }
            Session[CartSessionKey] = cart;
            return RedirectToAction("Index");
        }

        // POST: /Cart/Remove
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Remove(int productId)
        {
            var cart = GetCart();
            cart.Items.RemoveAll(x => x.ProductId == productId);
            Session[CartSessionKey] = cart;
            return RedirectToAction("Index");
        }

        // GET: /Cart/Checkout
        [Authorize]
        public ActionResult Checkout()
        {
            var cart = GetCart();
            if (cart.TotalQty == 0) return RedirectToAction("Index");
            var vm = new CheckoutViewModel { Cart = cart };
            return View(vm);
        }

        // POST: /Cart/Checkout
        [HttpPost]
        [Authorize]
        [ValidateAntiForgeryToken]
        public ActionResult Checkout(CheckoutViewModel model)
        {
            if (!ModelState.IsValid)
            {
                model.Cart = GetCart();
                return View(model);
            }
            // TODO: 建立訂單、清空購物車
            Session.Remove(CartSessionKey);
            return RedirectToAction("Complete");
        }

        // GET: /Cart/Complete
        public ActionResult Complete()
        {
            return View();
        }
    }
}
